<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.26">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Hande Acar Kirit, PhD OU Health Campus, IRCF Bioinformatics Division">
<meta name="dcterms.date" content="2025-12-09">

<title>Cell Ranger for scRNA‑seq</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-587c61ba64f3a5504c4d52d930310e48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-07cabb20591597c73cd7c3898511ecc8.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<meta property="og:title" content="Cell Ranger for scRNA‑seq">
<meta property="og:description" content="">
<meta property="og:image" content="https://EleanaCab.github.io/scRNAseq-wshp/days-content/day-1-resources/images/workflow.webp">
</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title">Cell Ranger for scRNA‑seq</h1>
        </a>     
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
      <a href="../../index.html" class="sidebar-logo-link">
      <img src="../../images/logo/logo_2.png" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none">
      <img src="../../images/logo/logo_2.png" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none">
      </a>
      <div class="sidebar-tools-main">
    <a href="https://github.com/EleanaCab/scRNAseq-wshp/tree/main" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
</div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false"><i class="bi bi-house" role="img">
</i> 
 <span class="menu-text">Course information</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshop-overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshop-team.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Team</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshop-prereqs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prerequisites</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../workshop-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Data Summary</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false"><i class="bi bi-book" role="img">
</i> 
 <span class="menu-text">Pre-reading</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../pre-readings/day-2-pre-reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../pre-readings/day-3-pre-reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 3</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false"><i class="bi bi-pencil" role="img">
</i> 
 <span class="menu-text">Daily materials</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../days-content/day-1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../days-content/day-2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../days-content/day-3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Day 3</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false"><i class="bi bi-clipboard2-plus" role="img">
</i> 
 <span class="menu-text">Supplemental notes</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../supplemental-material/additional-links.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Additional Resources</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../supplemental-material/install-packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install Packages Script</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">1. Introduction</a>
  <ul class="collapse">
  <li><a href="#from-fastq-file-to-count-matrix" id="toc-from-fastq-file-to-count-matrix" class="nav-link" data-scroll-target="#from-fastq-file-to-count-matrix">From FASTQ File to Count Matrix</a></li>
  <li><a href="#prerequisites" id="toc-prerequisites" class="nav-link" data-scroll-target="#prerequisites">Prerequisites</a></li>
  <li><a href="#connecting-to-oscer-interactively" id="toc-connecting-to-oscer-interactively" class="nav-link" data-scroll-target="#connecting-to-oscer-interactively">Connecting to OSCER interactively</a></li>
  </ul></li>
  <li><a href="#genome-references-prebuilt-vs.-custom" id="toc-genome-references-prebuilt-vs.-custom" class="nav-link" data-scroll-target="#genome-references-prebuilt-vs.-custom">2. Genome References: prebuilt vs.&nbsp;custom</a>
  <ul class="collapse">
  <li><a href="#option-a-use-a-prebuilt-reference-easier-if-exist" id="toc-option-a-use-a-prebuilt-reference-easier-if-exist" class="nav-link" data-scroll-target="#option-a-use-a-prebuilt-reference-easier-if-exist">Option A: Use a prebuilt reference (easier if exist)</a></li>
  <li><a href="#option-b-build-a-custom-reference-with-mkref" id="toc-option-b-build-a-custom-reference-with-mkref" class="nav-link" data-scroll-target="#option-b-build-a-custom-reference-with-mkref">Option B: Build a custom reference with <code>mkref</code></a></li>
  </ul></li>
  <li><a href="#running-cellranger-count" id="toc-running-cellranger-count" class="nav-link" data-scroll-target="#running-cellranger-count">3. Running <code>cellranger count</code></a></li>
  <li><a href="#outputs-youll-use" id="toc-outputs-youll-use" class="nav-link" data-scroll-target="#outputs-youll-use">4. Outputs you’ll use</a></li>
  <li><a href="#understanding-the-qc-summary-web_summary.html" id="toc-understanding-the-qc-summary-web_summary.html" class="nav-link" data-scroll-target="#understanding-the-qc-summary-web_summary.html">5. Understanding the QC summary <code>web_summary.html</code></a></li>
  <li><a href="#running-cellranger-over-oscer-using-scheduler" id="toc-running-cellranger-over-oscer-using-scheduler" class="nav-link" data-scroll-target="#running-cellranger-over-oscer-using-scheduler">6. Running CellRanger over OSCER using scheduler</a></li>
  <li><a href="#other-single-cell-technologies" id="toc-other-single-cell-technologies" class="nav-link" data-scroll-target="#other-single-cell-technologies">7. Other single cell technologies:</a>
  <ul class="collapse">
  <li><a href="#parse-biosciences" id="toc-parse-biosciences" class="nav-link" data-scroll-target="#parse-biosciences">Parse Biosciences</a></li>
  <li><a href="#illumina-single-cell-3-rna-prep" id="toc-illumina-single-cell-3-rna-prep" class="nav-link" data-scroll-target="#illumina-single-cell-3-rna-prep">Illumina Single Cell 3’ RNA Prep</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Cell Ranger for scRNA‑seq</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Hande Acar Kirit, PhD<br>
OU Health Campus, IRCF Bioinformatics Division </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 9, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<blockquote class="blockquote">
<p>This session walks you through Cell Ranger software for 10x Genomics Gene Expression data. It covers prerequisites, genome references (including making a human reference with <code>mkref</code>), preparing FASTQs, running <code>cellranger count</code>, and interpreting key outputs. Commands are provided as Bash chunks you can run on Linux. Adjust paths and resource flags to match your system.</p>
</blockquote>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">1. Introduction</h2>
<p>Single cell RNA-seq workflows are typically broken into three stages: primary, secondary, and tertiary analysis. Primary analysis handles the initial processing of raw sequencing output, including assigning reads to samples based on their barcode indices and trimming the reads. Then the reads are aligned to a reference and a quantitative measures of gene expression is generated as a count matrix. Secondary analysis involves dimensionality reduction and clustering, and finally tertiary analysis centers on interpreting the data biologically, most of the time through annotating the cells and performing differential gene expression testing and enrichment analyses such as GO or gene set enrichment.</p>
<p>There are several analysis pipelines developed to perform the primary analysis step of the single cell data analysis and your choice sometimes depend on the method by which the library was generated:</p>
<ul>
<li><p><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md">STAR solo</a>: This tool is built into the general purpose STAR aligner and generates outputs very similar to Cell Ranger minus the cloupe file and the QC report (see below). The advantage over Cell Ranger is that it is much less computationally intensive and will run with lower memory requirements in a shorter time.</p></li>
<li><p><a href="https://salmon.readthedocs.io/en/latest/alevin.html">Alevin</a>: This tool is based on the popular Salmon tool for bulk RNAseq feature counting and performs pseudoalignment, and therefore, it is significantly faster. Currently alevin supports the following single-cell protocols: Drop-seq, 10x-Chromium v1/2/3, inDropV2, CELSeq 1/2, Quartz-Seq2, sci-RNA-seq3.</p></li>
<li><p><a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/getting-started/cr-what-is-cell-ranger">CellRanger</a>: A set of analysis pipelines developed by 10X (Cell Ranger actually uses STAR under the hood). This software does not only perform the alignment and feature counting, but will also call cells (i.e.&nbsp;filter the raw matrix to remove droplets that do not contain cells), generate a useful report in html format, which will provide some QC metrics and an initial look at the data, and generate a “cloupe” file, which can be opened using the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser">10x Loupe Browser</a> software to further explore the data.</p>
<p>CellRanger is computationally very intensive because it performs both primary and secondary analysis steps. As a result, it cannot be run on a standard laptop or desktop computer; you will need access to a high-performance computing (HPC) cluster. Since we are working with 10x Chromium–derived data, we will use CellRanger for this workshop. For detailed instructions, please refer to the CellRanger documentation on the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10x Genomics website</a>.</p>
<p>However, because secondary analysis requires careful tuning of parameters specific to each dataset, we do not recommend relying solely on CellRanger’s automated results. Instead, we will carry out the secondary and tertiary analysis steps using relevant R packages on Day 2 and 3 of the workshop.</p>
<p>You can also run 10x Genomics single cell pipelines with <a href="https://www.10xgenomics.com/products/cloud-analysis">10x Genomics Cloud Analysis</a>, creating a free account. he analysis pipeline has some limitations, but usually performs well. You need to download your output within 3 months.</p></li>
</ul>
<p><br></p>
<section id="from-fastq-file-to-count-matrix" class="level3">
<h3 class="anchored" data-anchor-id="from-fastq-file-to-count-matrix">From FASTQ File to Count Matrix</h3>
<p><img src="images/workflow.webp" class="img-fluid" style="width:100.0%" data-fig-cap="Workflow"></p>
<p>In bulk RNA-seq experiment, we usually run some QC checks on fastq files containing reads, then map the reads to a reference genome, and finally generate a count matrix for every gene across each sample. In single cell RNA-seq experiment, similarly, we generate a count matrix that represents the number of counts for every gene across each profiled cell in the experiment.</p>
<p>For scRNA-seq data processing, we need these three components before we can start with the secondary and tertiary analysis:<br>
- Sparse Matrix (MTX file) represents the count data in a compact format.<br>
- Cell Metadata containing barcode information.<br>
- Gene List that includes gene names and IDs.</p>
</section>
<section id="prerequisites" class="level3">
<h3 class="anchored" data-anchor-id="prerequisites">Prerequisites</h3>
<p>We will be looking at the <code>count</code> tool of the CellRanger software to align the sequencing reads against the reference genome of your organism, quantify gene expression and call cells. To generate gene (feature) counts you will need:</p>
<ul>
<li><strong>Linux</strong> environment (workstation, HPC, or cloud) with enough RAM/CPU and disk space: we will be using <a href="https://www.ou.edu/oscer/getting_started/getting_started_hpc_intro">OSCER HPC</a>.<br>
</li>
<li><strong>CellRanger (v8+ recommended; v9 current)</strong>: it is already installed on OSCER.<br>
</li>
<li><strong>FASTQs</strong> demultiplexed from the sequencer (see “FASTQ naming”): we will provide these.<br>
</li>
<li>A <strong>reference transcriptome</strong>: either <strong>prebuilt on 10x website</strong> or a <strong>custom reference</strong> you build with <code>mkref</code>.</li>
</ul>
</section>
<section id="connecting-to-oscer-interactively" class="level3">
<h3 class="anchored" data-anchor-id="connecting-to-oscer-interactively">Connecting to OSCER interactively</h3>
<p>To perform the next steps, we will connect to OSCER interactively, since we do not want to overload the login node and risk receiving a dreaded email from <a href="https://www.ou.edu/oscer/people">Dr.&nbsp;Henry Neeman, Director of the OU Supercomputing Center for Education &amp; Research</a>.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--partition</span><span class="op">=</span>sooner_test <span class="at">--container</span><span class="op">=</span>el9hw <span class="at">--cpus-per-task</span><span class="op">=</span>2 <span class="at">--mem</span><span class="op">=</span>8G <span class="at">--pty</span> bash <span class="at">-i</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><br></p>
</section>
</section>
<section id="genome-references-prebuilt-vs.-custom" class="level2">
<h2 class="anchored" data-anchor-id="genome-references-prebuilt-vs.-custom">2. Genome References: prebuilt vs.&nbsp;custom</h2>
<section id="option-a-use-a-prebuilt-reference-easier-if-exist" class="level3">
<h3 class="anchored" data-anchor-id="option-a-use-a-prebuilt-reference-easier-if-exist">Option A: Use a prebuilt reference (easier if exist)</h3>
<p>10x publishes prebuilt reference packages for several organisms on its <a href="https://www.10xgenomics.com/support/software/cell-ranger/downloads#reference-downloads">website</a>. You can simply download the latest prebuilt reference (e.g., for mouse, refdata-gex-mm10-2020-A) following the script provided below and point <code>--transcriptome</code> to that folder. For a detailed script for how that reference was built, you can follow <a href="https://www.10xgenomics.com/support/software/cell-ranger/downloads/cr-ref-build-steps">the build steps</a> on the 10 website for the references.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Downloading refdata-gex-mm10-2020-A.tar.gz from 10x website</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/refs</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/refs</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> <span class="st">"https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-mm10-2020-A.tar.gz"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xzvf</span> refdata-gex-mm10-2020-A.tar.gz</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-1</span> ~/refs/refdata-gex-mm10-2020-A</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="option-b-build-a-custom-reference-with-mkref" class="level3">
<h3 class="anchored" data-anchor-id="option-b-build-a-custom-reference-with-mkref">Option B: Build a custom reference with <code>mkref</code></h3>
<p>Build a custom reference if you need non‑standard genomes, extra genes (e.g., reporters), or to align with a specific <a href="https://useast.ensembl.org/index.html">Ensembl/GENCODE</a> release. You need the reference genome for your organism in FASTA format and the transcript annotation file in GTF format. As an example, see the code below for the human reference genome release 110 on Ensembl. You can also check the instructions given on 10X website to <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/inputs/cr-3p-references">build custom reference</a> for more details.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: set your Ensembl release and target directory</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="va">REFDIR</span><span class="op">=</span>/path/to/referenceGenome <span class="co"># change the directory as you prefer</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">$REFDIR</span><span class="st">"</span> <span class="kw">&amp;&amp;</span> <span class="bu">cd</span> <span class="st">"</span><span class="va">$REFDIR</span><span class="st">"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Update the URLs below to the current Ensembl release if needed</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> Homo_sapiens.GRCh38.110.gtf.gz   https://ftp.ensembl.org/pub/release-110/gtf/homo_sapiens/Homo_sapiens.GRCh38.110.gtf.gz</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Human genome FASTA (primary assembly)</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz https://ftp.ensembl.org/pub/release-110/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Decompress for mkref/mkgtf since mkref does not work with files compressed with gzip</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="at">-f</span> Homo_sapiens.GRCh38.110.gtf.gz</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> <span class="at">-f</span> Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><strong>GTF (Gene Transfer Format)</strong></p>
<p><img src="images/GTFfile2.png" class="img-fluid" style="width:100.0%" data-fig-cap="GTF file format"></p>
<p><img src="images/GTF_columns.png" class="img-fluid" style="width:80.0%" data-fig-cap="GTF file format"></p>
<ul>
<li>Used for genome annotation.<br>
</li>
<li>Importantly for Cell Ranger Count, only features labelled as exon (column 3) will be considered for counting signal in genes.<br>
</li>
<li>Many genomes label mitochondrial genes with CDS and not exon so these must be updated.</li>
</ul>
<p><strong>Optional but recommended:</strong> 10x recommends filtering the GTF file so that it contains only gene categories of interest by using the <code>cellranger mkgtf</code> tool. Which genes to filter depends on your research question. Note that Cell Ranger does not use reads that map to multiple genes for expression (UMI) counting, so having overlapping genes in your reference will result in ignored reads. This filtering is recommended to avoid multi‑mapping from non‑polyA/non‑coding features. 10x demonstrates this with <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/inputs/cr-3p-references#filter-with-mkgtf-f4a9b6"><code>cellranger mkgtf</code></a>:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter the gtf file to contain only protein coding exons:</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cellranger mkgtf input.gtf output.gtf --attribute=key:allowable_value</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> mkgtf <span class="dt">\</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  Homo_sapiens.GRCh38.110.gtf <span class="dt">\</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  Homo_sapiens.GRCh38.110.protein_coding.gtf <span class="dt">\</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--attribute</span><span class="op">=</span>gene_biotype:protein_coding</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Other attributes that can be used for filtering in pre-built 10x Genomics references include:</p>
<ul>
<li>Protein-coding genes (<code>--attribute=gene_biotype:protein_coding</code>)<br>
</li>
<li>Long intergenic noncoding RNAs (<code>--attribute=gene_biotype:lncRNA</code>)<br>
</li>
<li>Antisense (<code>--attribute=gene_biotype:antisense</code>)<br>
</li>
<li>Pseudogenes (<code>--attribute=gene_biotype:pseudogene</code>)<br>
</li>
<li>V(D)J germline genes, for example:<br>
<code>--attribute=gene_biotype:IG_V_gene</code><br>
<code>--attribute=gene_biotype:TR_V_gene</code></li>
</ul>
<p>Now we can create the Cell Ranger reference (this will produce a folder you pass to <code>--transcriptome</code> of the <code>cellranger count</code>).</p>
<p><strong>But we will not, because it takes too long:</strong></p>
<p><strong>System requirements:</strong>. Indexing a typical human 3Gb FASTA file often takes up to 8 core hours and requires 32 GB of memory.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> mkref <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--nthreads</span><span class="op">=</span>{CPUS} <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genome</span><span class="op">=</span>/path/to/output/genome/index <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--fasta</span><span class="op">=</span>/path/to/GenomeFASTA/Homo_sapiens.GRCh38.dna.primary_assembly.fa <span class="dt">\</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genes</span><span class="op">=</span>/path/to/GenomeGTF/Homo_sapiens.GRCh38.110.protein_coding.gtf</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>About <strong>“include introns”</strong>: Since Cell Ranger <strong>v7</strong>, intronic reads are <strong>included by default</strong> for whole‑transcriptome GEX, which makes a separate “pre‑mRNA” reference unnecessary for snRNA‑seq. Use <code>--include-introns=false</code> to count exonic reads only if you are running snRNA-seq.</p>
</div>
</div>
</section>
</section>
<section id="running-cellranger-count" class="level2">
<h2 class="anchored" data-anchor-id="running-cellranger-count">3. Running <code>cellranger count</code></h2>
<p><strong>File naming:</strong> Cell Ranger expects the Illumina convention: <code>Sample_S1_L00{lane}_{I1|I2|R1|R2}_001.fastq.gz</code><br>
Lane‑less names like <code>Sample_S1_R1_001.fastq.gz</code> are also accepted (v4+).</p>
<p>Below is a template for a single‑sample run. For more information about <code>cellranger count</code> function check <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/tutorials/cr-tutorial-ct">10x website</a>.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Resources (adjust to your node/request)</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="va">CORES</span><span class="op">=</span>64</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="va">MEM_GB</span><span class="op">=</span>128</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Run</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> count <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--id</span><span class="op">=</span>{OUTPUT_SAMPLE_NAME} <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--transcriptome</span><span class="op">=</span>/path/to/output/genome/index <span class="dt">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--fastqs</span><span class="op">=</span>/path/to/fastqs <span class="dt">\</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sample</span><span class="op">=</span>{name-of-sample-as-in-fastq-files} <span class="dt">\</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">--chemistry</span><span class="op">=</span>auto <span class="dt">\</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">--localcores</span><span class="op">=</span><span class="st">"</span><span class="va">$CORES</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">--localmem</span><span class="op">=</span><span class="st">"</span><span class="va">$MEM_GB</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">--create-bam</span><span class="op">=</span>true</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><strong>Notes</strong>:</p>
<ul>
<li>This script will process all fastqs in <code>--fastqs</code> dir unless <code>--sample</code> is used; if multiple samples share a dir, use <code>--sample</code> to restrict.<br>
</li>
<li><code>--chemistry=auto</code> is the default and works for 3′ vs 5′ kits in most cases. Override only if the kit is unusual.</li>
</ul>
</section>
<section id="outputs-youll-use" class="level2">
<h2 class="anchored" data-anchor-id="outputs-youll-use">4. Outputs you’ll use</h2>
<p>Cell Ranger will create a single results folder for each sample names. Each folder will be named according to the <code>--id</code> option in the command.</p>
<p><img src="images/output.png" class="img-fluid" style="width:30.0%" data-fig-cap="output tree"></p>
<p><br></p>
<p>Each run creates <code>&lt;ID&gt;/outs/</code> with (key items):</p>
<p><img src="images/output_out.png" class="img-fluid" style="width:50.0%" data-fig-cap="output out tree"></p>
<ul>
<li><code>analysis</code> : The results of clustering and differential expression analysis on the clusters. These are used in the web_summary.html report.<br>
</li>
<li><code>cloupe.cloupe</code> : a cloupe file that can be opened in Loupe Browser for quick inspection<br>
</li>
<li><code>filtered_feature_bc_matrix</code> : The filtered count matrix directory<br>
</li>
<li><code>filtered_feature_bc_matrix.h5</code> : The filtered count matrix as an HDF5 file<br>
</li>
<li><code>metrics_summary.csv</code> : summary metrics from the analysis<br>
</li>
<li><code>molecule_info.h5</code> : per-molecule read information as an HDF5 file<br>
</li>
<li><code>possorted_genome_bam.bam</code> : The aligned reads in bam format<br>
</li>
<li><code>possorted_genome_bam.bam.bai</code> : The bam index<br>
</li>
<li><code>raw_feature_bc_matrix</code> : The raw count matrix directory<br>
</li>
<li><code>raw_feature_bc_matrix.h5</code> : The raw count matrix as an HDF5 file<br>
</li>
<li><code>web_summary.html</code> : interactive QC summary</li>
</ul>
<p><br></p>
<p>The two count matrix directories each contain 3 files:<br>
<img src="images/output_out_matrix.png" class="img-fluid" style="width:30.0%" data-fig-cap="output out tree"></p>
<ul>
<li><code>barcodes.tsv.gz</code> : The cell barcodes detected; these correspond to the columns of the count matrix<br>
</li>
<li><code>features.tsv.gz</code> : The features detected. In this cases gene ids. These correspond to the rows of the count matrix.<br>
</li>
<li><code>matrix.mtx.gz</code> : the count of unique UMIs for each gene in each cell.</li>
</ul>
<p>The count matrix directories and their corresponding HDF5 files contain the same information, they are just alternative formats for use depending on the tools that are used for analysis.</p>
<p>The filtered count matrix only contains droplets that have been called as cells by CellRanger.</p>
</section>
<section id="understanding-the-qc-summary-web_summary.html" class="level2">
<h2 class="anchored" data-anchor-id="understanding-the-qc-summary-web_summary.html">5. Understanding the QC summary <code>web_summary.html</code></h2>
<p>The cellranger count pipeline outputs an interactive summary HTML file named <code>web_summary.html</code> that contains summary metrics and automated secondary analysis results. It has Summary and Gene Expression tabs. For a general intro you can read <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-web-summary-count">this link here</a>.</p>
<p><img src="images/web_summary_1.png" class="img-fluid" style="width:100.0%" data-fig-cap="output out tree"></p>
<p>Click the <code>?</code> icons next to each section title to display information about the secondary analyses shown in the dashboard.</p>
<p>The Barcode Rank Plot is an interactive plot that shows all barcodes detected in an experiment, ranked from highest to lowest UMI count. It is useful for understanding <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/algorithms-overview/cr-gex-algorithm#cell_calling">Cell Ranger’s cell calling algorithm</a> to gain insight into your sample quality.</p>
<p>Below is a Barcode Rank Plot from a good sample:</p>
<p><img src="images/BarcodeRankPlot.png" class="img-fluid" style="width:80.0%" data-fig-cap="output out tree"></p>
<p>And here are some not so good ones: <img src="images/badsample1.png" class="img-fluid" style="width:80.0%" data-fig-cap="output out tree"></p>
<p><img src="images/badsample2.png" class="img-fluid" style="width:80.0%" data-fig-cap="output out tree"></p>
<p><img src="images/badsample3.png" class="img-fluid" style="width:80.0%" data-fig-cap="output out tree"></p>
<p>For a more detailed explanation on the interpretation of web summary files see <a href="https://cdn.10xgenomics.com/image/upload/v1660261286/support-documents/CG000329_TechnicalNote_InterpretingCellRangerWebSummaryFiles_RevA.pdf">this</a> info sheet from 10X.</p>
</section>
<section id="running-cellranger-over-oscer-using-scheduler" class="level2">
<h2 class="anchored" data-anchor-id="running-cellranger-over-oscer-using-scheduler">6. Running CellRanger over OSCER using scheduler</h2>
<p>Connecting to OSCER:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Log into OSCER HPC using your username and password</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@sooner.oscer.ou.edu</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># activate the el9 container so that modules can be loaded</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="ex">el9</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>You should never run a program over the login node, that may cause the login node to crash and cause your and other researchers’ runs to fail. We are doing this only to verify proper installation and functioning of CellRanger on OSCER HPC. It’s always a good idea to check every component of your run before you put your sbatch file together.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the cell ranger module (if you cannot remember versions, run &lt;module spider cellranger&gt; and it will list installed software with similar names)</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load CellRanger/8.0.1</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Check Cell Ranger is activated</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> <span class="at">--version</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># To see the help</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> <span class="at">--help</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>To run a software on OSCER we need to submit our jobs over a workload manager called <a href="https://slurm.schedmd.com/documentation.html">SLURM</a>.</p>
<section id="batch-jobs" class="level4">
<h4 class="anchored" data-anchor-id="batch-jobs">Batch Jobs</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/bash -l</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --job-name=CR_single</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --partition=el9_test</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --container=el9hw</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=64</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=128G</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=24:00:00</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --chdir=/path/to/your/output/</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">#===============================================================================</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load CellRanger/8.0.1</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>/path/to/your/output/<span class="op">&gt;</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="va">s</span><span class="op">=&lt;</span>give <span class="ex">your</span> sample name here<span class="op">&gt;</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"* </span><span class="va">$(</span><span class="fu">date</span><span class="va">)</span><span class="st"> | START sample: </span><span class="va">$s</span><span class="st">"</span> <span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.out"</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a><span class="ex">cellranger</span> count <span class="dt">\</span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>        <span class="at">--id</span> <span class="st">"</span><span class="va">$s</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a>        <span class="at">--sample</span> <span class="st">"</span><span class="va">$s</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastqs</span> <span class="st">"/path/to/fastqs"</span> <span class="dt">\</span></span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">--transcriptome</span> <span class="st">"/path/to/genome/index"</span> <span class="dt">\</span></span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">--localcores</span> 64 <span class="dt">\</span></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">--localmem</span> 128 <span class="dt">\</span></span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">--create-bam</span> true <span class="dt">\</span></span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">--disable-ui</span> <span class="dt">\</span></span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">--output-dir</span> <span class="st">"/path/to/your/output/</span><span class="va">${s}</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>        <span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.out"</span> <span class="dv">2</span><span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.err"</span></span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"**** </span><span class="va">$(</span><span class="fu">date</span><span class="va">)</span><span class="st"> | DONE sample: </span><span class="va">$s</span><span class="st">"</span> <span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.out"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><strong>Resource control</strong> Cell Ranger can use most available cores/RAM; constrain with <code>--localcores</code> and <code>--localmem</code> to match your scheduler request.<br>
</li>
<li><strong>Disk space</strong> Keep ~3–4× input FASTQ size free for temporary files and outputs.<br>
</li>
<li><strong>FASTQ selection</strong> If a directory holds multiple samples, use <code>--sample</code> to restrict to the intended prefix.</li>
<li><strong>snRNA‑seq</strong> Introns are counted by default (v7+); you do <strong>not</strong> need a pre‑mRNA reference.</li>
<li><strong>Chemistry detection</strong> By default the assay configuration is detected automatically, leave it out unless you have reason to override; if you must, set <code>--chemistry</code> to one of the accepted values (e.g., <code>threeprime</code>, <code>fiveprime</code>, etc.).</li>
<li>Here is the link for more on <a href="https://www.10xgenomics.com/support/software/cell-ranger/latest/resources/cr-command-line-arguments">Cell Ranger Command Line Arguments.</a></li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Pathway to data for this workshop
</div>
</div>
<div class="callout-body-container callout-body">
<p>fastq files are under: <code>"/ourdisk/hpc/iicomicswshp/dont_archive/omics_workshop_2025/mouse_kidney_all_cells_raw_data/"</code></p>
<p>reference is under: <code>"/ourdisk/hpc/iicomicswshp/dont_archive/master_folder/omics_class_scripts/data/sc/refs/refdata-gex-mm10-2020-A"</code></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Running samples serially
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --job-name=CR_serial</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --partition=el9_test</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --container=el9hw</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --nodes=1</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task=64</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=128G</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=24:00:00</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --chdir=/path/to/your/output/</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load CellRanger/8.0.1</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> <span class="op">&lt;</span>/path/to/your/output/<span class="op">&gt;</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="co"># this will run samples one after the other (serially):</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="co"># make sure there is no empty spaces or empty lines in the ids.txt</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="va">IDS</span><span class="op">=</span><span class="st">"/ourdisk/hpc/iicomicswshp/dont_archive/omics_workshop_2025/mouse_kidney_all_cells_raw_data/ids.txt"</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> s <span class="kw">in</span> <span class="va">$(</span><span class="fu">cat</span> <span class="st">"</span><span class="va">$IDS</span><span class="st">"</span><span class="va">)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"* </span><span class="va">$(</span><span class="fu">date</span><span class="va">)</span><span class="st"> | START sample: </span><span class="va">$s</span><span class="st">"</span> <span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.out"</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>    <span class="ex">cellranger</span> count <span class="dt">\</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>        <span class="at">--id</span> <span class="st">"</span><span class="va">$s</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">--sample</span> <span class="st">"</span><span class="va">$s</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastqs</span> <span class="st">"/path/to/fastqs"</span> <span class="dt">\</span></span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">--transcriptome</span> <span class="st">"/path/to/genome/index"</span> <span class="dt">\</span></span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">--localcores</span> 64 <span class="dt">\</span></span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">--localmem</span> 128 <span class="dt">\</span></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">--create-bam</span> true <span class="dt">\</span></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">--disable-ui</span> <span class="dt">\</span></span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>        <span class="at">--output-dir</span> <span class="st">"/path/to/your/output/</span><span class="va">${s}</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>        <span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.out"</span> <span class="dv">2</span><span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.err"</span></span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"**** </span><span class="va">$(</span><span class="fu">date</span><span class="va">)</span><span class="st"> | DONE sample: </span><span class="va">$s</span><span class="st">"</span> <span class="op">&gt;&gt;</span> <span class="st">"</span><span class="va">${s}</span><span class="st">.out"</span></span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"All samples finished."</span></span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="other-single-cell-technologies" class="level2">
<h2 class="anchored" data-anchor-id="other-single-cell-technologies">7. Other single cell technologies:</h2>
<section id="parse-biosciences" class="level3">
<h3 class="anchored" data-anchor-id="parse-biosciences">Parse Biosciences</h3>
<p><a href="https://www.parsebiosciences.com/blog/getting-started-with-scrna-seq-post-sequencing-data-analysis/">Parse Biosciences</a> offers an alternative strategy to 10X technology using combinatorial in-situ barcoding. If you have data from Parse Biosciences you can analyze the data on their analysis interface <a href="https://biomage-auth-production-242905224710.auth.eu-west-1.amazoncognito.com/login?redirect_uri=https%3A%2F%2Fapp.trailmaker.parsebiosciences.com%2Fredirect&amp;response_type=code&amp;client_id=665t39tl77h7q94f6ssvojmh77&amp;identity_provider=COGNITO&amp;scope=aws.cognito.signin.user.admin%20email%20openid%20phone%20profile&amp;state=hOwtQBIe3LtQdH3bWHRUgEqAq41ryJYo&amp;code_challenge=Gyuwkbq9e8-VuCsVTrdGG_3Vsg3q6PldV2EJYnfBam0&amp;code_challenge_method=S256">Trailmaker</a> for free (with limitations). See <a href="https://youtu.be/9aVNh6kCTOY?si=ag_GSJOAcb5u6IE0">this video</a> to learn how to.</p>
</section>
<section id="illumina-single-cell-3-rna-prep" class="level3">
<h3 class="anchored" data-anchor-id="illumina-single-cell-3-rna-prep">Illumina Single Cell 3’ RNA Prep</h3>
<p>Illumina offers a droplet-based strategy as an alternative. See their <a href="https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/single-cell-rna-prep.html">website</a> for more information on the approach. If you have data from Illumina droplet-based technology, you can analyze the data on their analysis interface <a href="https://basespace.illumina.com/apps/18275257">BaseSpace DRAGEN Single Cell RNA app.</a></p>
<p><br></p>
<hr>
<p><strong>Acknowledgements &amp; references</strong></p>
<ul>
<li><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview">10x Genomics</a>: official tutorials for <code>mkref</code> and <code>count</code>, reference build notes, FASTQ specs, algorithm details, and system requirements.<br>
</li>
<li><a href="https://bioinformatics-core-shared-training.github.io/UnivCambridge_ScRnaSeq_Nov2021/Markdowns/03_CellRanger.html">University of Cambridge Bioinformatics Training (Nov 2021)</a><br>
</li>
<li><a href="https://rockefelleruniversity.github.io/scRNA-seq/presentations/slides/Session1.html#1">Rockefeller University, Bioinformatics Resource Centre, Single-cell RNA sequencing workshop</a></li>
<li>workflow image is taken from <a href="https://www.parsebiosciences.com/blog/getting-started-with-scrna-seq-post-sequencing-data-analysis/">Parse Biosciences website</a></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/EleanaCab\.github\.io\/scRNAseq-wshp\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>